ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/CSJ.hs

```diff
@@ -0,0 +1,134 @@
+      ]
+    ]
+
+-- | Test of the “happy path” scenario of ChainSync Jumping (CSJ).
+--
+-- This test features one chain (ie. a block tree that is only trunk) and only
+-- honest peers and syncs the chain in question with CSJ enabled. What we expect
+-- to observe is that one of the honest peers becomes the dynamo while the
+-- others become jumpers. Because the jumpers will agree to all the jumps, the
+-- whole syncing should happen with CSJ without objectors.
+--
+-- The final property is that headers should only ever be downloaded once and
+-- only from one peer (the dynamo). This is true except when almost caught-up:
+-- when the dynamo is caught-up, it gets disengaged and one of the jumpers takes
+-- its place and starts serving headers. This might lead to duplication of
+-- headers, but only in a window of @jumpSize@ slots near the tip of the chain.
+--
+-- The boolean differentiates between “synchronous” and “asynchronous”
+-- scenarios. In a synchronous scenario, all the honest peers have the same
+-- schedule: they serve the chain exactly in the same way. In the asynchronous
+-- scenario, a random schedule is generated for each peer (but they still serve
+-- the same chain).
+prop_happyPath :: Bool -> Property
+prop_happyPath synchronized =
+  forAllGenesisTest
+    ( do
+        gt <- genChains $ pure 0
+        honest <- genHonestSchedule gt
+        numOthers <- choose (1, 3)
+        otherHonests <- if synchronized
+          then pure $ replicate numOthers honest
+          else replicateM numOthers (genHonestSchedule gt)
+        pure $ gt $> mkPeers honest otherHonests
+    )
+    ( defaultSchedulerConfig
+      { scEnableCSJ = True
+      , scEnableLoE = True
+      , scEnableLoP = True
+      }
+    )
+    ( -- NOTE: Shrinking makes the tests fail because the peers reject jumps
+      -- because their TP is G. This makes them into objectors and they then
+      -- start serving headers.
+      \_ _ -> []
+    )
+    ( \gt StateView{svTrace} ->
+        let
+          headerDownloadEvents =
+            mapMaybe
+              (\case
+                TraceChainSyncClientEvent pid (TraceDownloadedHeader hdr)
+                  | not (isNewerThanJumpSizeFromTip gt hdr)
+                  -> Just (pid, hdr)
+                _ -> Nothing
+              )
+              svTrace
+          receivedHeadersOnlyOnce = length (nub $ snd <$> headerDownloadEvents) == length headerDownloadEvents
+          -- NOTE: If all the headers are newer than jumpSize from the tip, then
+          -- 'headerDownloadEvents' is empty and the following condition would
+          -- violated if we used @==@.
+          receivedHeadersFromOnlyOnePeer = length (nubOrd $ fst <$> headerDownloadEvents) <= 1
+        in
+          tabulate ""
+            [ if headerDownloadEvents == []
+                then "All headers may be downloaded twice"
+                else "There exist headers that have to be downloaded exactly once"
```

Comment:
The meaning of `headerDownloadEvents` may require a comment.

> all headers have to be downloaded once 

We always expect some headers to be downloaded more than once. The first time is when the first dynamo serves them. It reaches the end of the chain, and it is disengaged. The next time will be when the next dynamo serves the headers since its last jump as a jumper. It reaches the end of the chain, and it is disengaged. And so on ...

Reply:
I tried to clarify the meaning of `headerDownloadEvents` in 980c35a986860c522273a236413dde44f0d4af15.